### ISSUE TITLE
sdc-login fails

### ISSUE THREAD

**Issue by sw00choi** [joyent/sdcadm#26]
Some VMs will be lost if any compute node in the Triton fails.
In this case, ```sdc-login``` to any VM in the headnode gets error.
```
/opt/smartdc/sdcadm/lib/sdcadm.js:720
                        vm.server_uuid].hostname;
                                       ^
TypeError: Cannot read property 'hostname' of undefined
    at Object.fillOutVmInsts [as func] (/opt/smartdc/sdcadm/lib/sdcadm.js:720:40)
    at Object._onImmediate (/opt/smartdc/sdcadm/node_modules/vasync/lib/vasync.js:213:20)
    at processImmediate [as _immediateCallback] (timers.js:330:15)
```

This error does not allow me to access any existing VM because of the missing VMs.

ctx.serverFromUuid[vm.server_uuid] in ```/opt/smartdc/sdcadm/lib/sdcadm.js``` gives me 'undefined' for the missing VM, so accessing the attributes to 'undefined' occurs error.
```
                // A state='provisioning' VM might not yet have a
                // 'server_uuid'.
                if (vm.server_uuid) {
                    inst.server = vm.server_uuid;
                    inst.hostname = ctx.serverFromUuid[
                        vm.server_uuid].hostname;                 // <-- error occurs here
                    inst.server_ip = ctx.serverAdminIpFromUuid[
                        vm.server_uuid];
                }
```

**Comment by askfongjojo** [comment]
Hi @sw00choi, you may have an older version of sdcadm since there is already a fix for this issue in August. Please update your sdcadm by doing:
```
sdcadm self-update -C dev --latest
```

**Comment by sw00choi** [comment]
Hi @askfongjojo.

The tested version information is:
```
- SmartOS (build: 20160923T025052Z)
[root@headnode (coal-1) ~]# sdcadm --version
sdcadm 1.12.0 (master-20161021T170133Z-gb4c58fe)
```

```sdcadm``` version indicates that the August patch is applied.

**Comment by kusor** [comment]
Hi @sw00choi, I've been unsuccessfully trying to reproduce the issue you mention here taking down CNs with VMs owned by the admin user.

Would it be possible to get the output of `sdc-cnapi /servers?extras=sysinfo,agents|json -Ha` and  `sdc-vmapi /vms?state=active\&owner_uuid=$(sdc-ufds search login=admin|json -Ha uuid)|json -H` in order to try to figure out what's going on with your setup?

Thanks!

**Comment by sw00choi** [comment]
Hi @kusor, I attached 2 files which contain the json responses of your commands.
[sdc-cnapi.txt](https://github.com/joyent/sdcadm/files/570432/sdc-cnapi.txt)
[sdc-vmapi.txt](https://github.com/joyent/sdcadm/files/570433/sdc-vmapi.txt)

Also, this might help you figure out which one is wrong.
I wrote some codes to get some error log like this:
```
# add try-catch block to log some error info.
try {
    inst.hostname = ctx.serverFromUuid[
    vm.server_uuid].hostname;
    inst.server_ip = ctx.serverAdminIpFromUuid[
    vm.server_uuid];
} catch(err) {
    self.log.error({err: err}, JSON.stringify(inst, null, 2));
}
```

Then, the error log looks like this:
```
[root@headnode (coal-1) ~]# cat /var/log/sdcadm/logs/1478220605731-077114-instances.log | bunyan -o short
00:50:06.995Z ERROR sdcadm/instances: (req_id=6aef01e1-f1ca-4b82-9dfb-8d45c6ce5242)
    {
      "type": "vm",
      "alias": "nat-fda8406c-fa27-425d-8380-a8c9bc9eb2ef",
      "version": "master-20160609T055533Z-ge6109fb",
      "instance": "6b3d030d-2812-4750-8846-ec61fecb55ed",
      "zonename": "6b3d030d-2812-4750-8846-ec61fecb55ed",
      "service": "nat",
      "image": "075d4bec-2e07-11e6-bd6d-5b0a73e44770",
      "state": "running",
      "server": "564df1e9-ec2f-5da0-302b-6b5148e7620b",
      "hostname": null,
      "server_ip": null
    }
    --
    TypeError: Cannot read property 'hostname' of undefined
        at Object.fillOutVmInsts [as func] (/opt/smartdc/sdcadm/lib/sdcadm.js:720:40)
        at Object._onImmediate (/opt/smartdc/sdcadm/node_modules/vasync/lib/vasync.js:213:20)
        at processImmediate [as _immediateCallback] (timers.js:330:15)
```

**Comment by kusor** [comment]
So, you have the vm `nat-fda8406c-fa27-425d-8380-a8c9bc9eb2ef` running but the server that vm is supposed to be hosted (`564df1e9-ec2f-5da0-302b-6b5148e7620b`) doesn't exist anymore, correct?

If that's the case, could you try to destroy the instance record from SAPI using `sdc-sapi /instances/6b3d030d-2812-4750-8846-ec61fecb55ed -X DELETE`?

I cannot imagine why this instance has been left around after server removal or why VMAPI still thinks it's running. Will open an issue to try to figure out that and another one for sdcadm to prevent this thing to disallow you using `sdcadm insts`.

**Comment by sw00choi** [comment]
Correct. The vm (```nat-fda8406c-fa27-425d-8380-a8c9bc9eb2ef```) can be found in adminui, but the CN (```564df1e9-ec2f-5da0-302b-6b5148e7620b```) for the vm does not exists anymore.

sdc-login was successful after ```sdc-sapi /instances/6b3d030d-2812-4750-8846-ec61fecb55ed -X DELETE```.

**Comment by kusor** [comment]
This is fixed by commit 68e25103c0
